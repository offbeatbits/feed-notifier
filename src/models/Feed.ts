import Parser from 'rss-parser';

export class Feed {
  url: string;

  private content: any;

  constructor(url: string) {
    this.url = url;
  }

  async fetch() {
    this.content = await new Parser().parseURL(this.url);
    return this;
  }

  getLatestPostDate() {
    return this.content.items[0].pubDate;
  }

  getLatestPostTitle() {
    return this.content.items[0].title;
  }

  getLatestPostUrl() {
    return this.content.items[0].link;
  }

  getLatestPostId() {
    return this.content.items[0].id;
  }
}
