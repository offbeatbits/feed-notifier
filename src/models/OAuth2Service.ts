import { OAuth2ServiceSettings } from "../types.d";

export abstract class OAuth2Service {
  key: string;

  secret: string;

  accessToken: string | undefined;

  accessTokenSecret: string | undefined;

  constructor({
    key,
    secret,
    accessToken,
    accessTokenSecret,
  }: OAuth2ServiceSettings) {
    this.key = key;
    this.secret = secret;
    this.accessToken = accessToken;
    this.accessTokenSecret = accessTokenSecret;
  }

  abstract post(content: string): void;
}
