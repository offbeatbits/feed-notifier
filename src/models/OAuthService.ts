import { OAuthServiceSettings } from '../types';

export class OAuthService {
  key: string;

  secret: string;

  accessToken: string;

  constructor(settings: OAuthServiceSettings) {
    this.key = settings.key;
    this.secret = settings.secret;
    this.accessToken = settings.accessToken || '';
  }
}
