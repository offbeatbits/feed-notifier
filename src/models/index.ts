export * from './Channel';
export * from './Feed';
export * from './OAuth2Service';
