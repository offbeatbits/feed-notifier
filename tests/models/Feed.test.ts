import { Feed } from "../../src/models/Feed";

jest.mock("rss-parser", () =>
  jest.fn().mockImplementation(() => ({
    parseURL: () =>
      Promise.resolve({
        content: {
          lastBuildDate: "mocked_last_build_date",
        },
        lastBuildDate: "mocked_last_build_date",
        items: [
          {
            pubDate: "mocked_last_item_pub_date",
            title: "mocked_last_item_title",
            link: "mocked_last_item_link",
          },
        ],
      }),
  }))
);

// eslint-disable-next-line jest/prefer-lowercase-title
describe("Feed model", () => {
  afterAll(async () => {
    await new Promise<void>((resolve) => setTimeout(() => resolve(), 500)); // avoid jest open handle error
  });

  const feedInstance = new Feed("sample_feed_url");

  const feedMethods = [
    "fetch",
    "getFeedUpdateDate",
    "getLatestPostDate",
    "getLatestPostTitle",
    "getLatestPostUrl",
  ];

  it("should be defined", () => {
    expect(Feed).toBeDefined();
  });

  it("should implement all methods", () => {
    feedMethods.forEach((method) => {
      expect((feedInstance as any)[method]).toBeDefined();
    });
  });

  it("should return feed update date", async () => {
    expect.assertions(1);
    const feed = await feedInstance.fetch();
    expect(feed.getFeedUpdateDate()).toMatchSnapshot();
  });

  it("should return latest post date", async () => {
    expect.assertions(1);
    const feed = await feedInstance.fetch();
    expect(feed.getLatestPostDate()).toMatchSnapshot();
  });

  it("should return latest post title", async () => {
    expect.assertions(1);
    const feed = await feedInstance.fetch();
    expect(feed.getLatestPostTitle()).toMatchSnapshot();
  });

  it("should return latest post URL", async () => {
    expect.assertions(1);
    const feed = await feedInstance.fetch();
    expect(feed.getLatestPostUrl()).toMatchSnapshot();
  });
});
