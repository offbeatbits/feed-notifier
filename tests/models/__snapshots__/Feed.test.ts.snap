// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feed model should return feed update date 1`] = `"mocked_last_build_date"`;

exports[`Feed model should return latest post URL 1`] = `"mocked_last_item_link"`;

exports[`Feed model should return latest post date 1`] = `"mocked_last_item_pub_date"`;

exports[`Feed model should return latest post title 1`] = `"mocked_last_item_title"`;
